evm/db/journal.py

```diff
@@ -1,226 +1,258 @@
         """
-        Returns all changes from the given checkpoint.  This includes all of
-        the changes from any subsequent checkpoints, giving precidence to
-        earlier checkpoints.
+        Returns all changes from the given changeset.  This includes all of
+        the changes from any subsequent changeset, giving precidence to
+        later changesets.
         """
-        idx = self.checkpoints.index(checkpoint_id)
+        if changeset_id not in self.journal_data:
+            raise KeyError("Unknown changeset: {0}".format(changeset_id))
 
-        # update the checkpoint list
-        checkpoint_ids = self.checkpoints[idx:]
-        self.checkpoints = self.checkpoints[:idx]
+        all_ids = tuple(self.journal_data.keys())
+        changeset_idx = all_ids.index(changeset_id)
+        changesets_to_pop = all_ids[changeset_idx:]
 
-        # we pull all of the checkpoints *after* the checkpoint we are
-        # reverting to and collapse them to a single set of keys that need to
-        # be reverted (giving precidence to earlier checkpoints).
-        revert_data = merge(*(
+        # we pull all of the changesets *after* the changeset we are
+        # reverting to and collapse them to a single set of keys (giving
+        # precedence to later changesets)
+        changeset_data = merge(*(
             self.journal_data.pop(c_id)
             for c_id
-            in reversed(checkpoint_ids)
+            in changesets_to_pop
         ))
 
-        return dict(revert_data.items())
+        return changeset_data
 
-    def commit_checkpoint(self, checkpoint_id):
+    def commit_changeset(self, changeset_id: uuid.UUID) -> Dict[bytes, bytes]:
         """
-        Collapses all changes for the givent checkpoint into the previous
-        checkpoint if it exists.
+        Collapses all changes for the given changeset into the previous
+        changesets if it exists.
         """
-        changes_to_merge = self.pop_checkpoint(checkpoint_id)
-        if self.checkpoints:
-            # we only have to merge the changes into the latest checkpoint if
+        changeset_data = self.pop_changeset(changeset_id)
+        if not self.is_empty():
+            # we only have to merge the changes into the latest changeset if
             # there is one.
             self.latest = merge(
-                changes_to_merge,
                 self.latest,
+                changeset_data,
             )
+        return changeset_data
+
+    #
+    # Database API
+    #
+    def get(self, key: bytes) -> bytes:
+        """
+        For key lookups we need to iterate through the changesets in reverse
+        order, returning from the first one in which the key is present.
+        """
+        # Ignored from mypy because of https://github.com/python/typeshed/issues/2078
+        for changeset_data in reversed(self.journal_data.values()):  # type: ignore
+            if key in changeset_data:
+                return changeset_data[key]
```

Comment:
Shouldn't this check that it's not returning `DELETED_ENTRY`?  Same issue I'm trying to understand here: https://github.com/ethereum/py-evm/pull/569/files/7a6d2e4b0a6a5349f09aaf693f1e94dd6bafe7ec#r184168693

Reply:
Well in case of the `Journal` (not the `JournalDB`) we may actually *want* to return the `DELETED_ENTRY` to give the higher order API the proper view to introspect. E.g. the `JournalDB` needs to know about the `DELETED_ENTRY` here in order to delete those entries from the underlying db. https://github.com/cburgdorf/py-evm/blob/7a6d2e4b0a6a5349f09aaf693f1e94dd6bafe7ec/evm/db/journal.py#L235

Reply:
This I don't mind.  Just wanted to make sure it was caught and handled in the `JournalDB`.